src/raft/raft.go

```diff
@@ -41,25 +46,344 @@ type ApplyMsg struct {
 // A Go object implementing a single Raft peer.
 //
 type Raft struct {
-	mu        sync.Mutex
+	mu          sync.RWMutex
+	role        uint64 //atomic
+	currentTerm int    //atomic counter
+
 	peers     []*labrpc.ClientEnd
 	persister *Persister
 	me        int // index into peers[]
 
+	majority int
+
+	electionTimeout  time.Duration
+	heartbeatTimeout time.Duration
+	leaderLease      time.Duration
+
+	heartBeatChan chan bool
+	lastContact   time.Time
+
+	killed bool
+	killCh chan bool
+
 	// Your data here.
 	// Look at the paper's Figure 2 for a description of what
 	// state a Raft server must maintain.
 
 }
 
+func (rf *Raft) isLeader() (isLeader bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isLeader = rf.role == 2
+	return
+}
+
+func (rf *Raft) isCandidate() (isCandidate bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isCandidate = rf.role == 1
+	return
+}
+
+func (rf *Raft) isFollower() (isFollower bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isFollower = rf.role == 0
+	return
+}
+
+func (rf *Raft) makeLeader() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 2
+	return
+}
+
+func (rf *Raft) makeCandidate() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 1
+	return
+}
+
+func (rf *Raft) makeFollower() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 0
+	return
+}
+
+func (rf *Raft) incTerm() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm++
+	return
+}
+
+func (rf *Raft) decTerm() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm--
+	return
+}
+
+func (rf *Raft) setTerm(term int) {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm = term
+	return
+}
+
+func (rf *Raft) getMajority() int {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	return rf.majority
+
+}
+
+func (rf *Raft) setLastContact() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.lastContact = time.Now()
+}
+
+func (rf *Raft) LastContact() time.Time {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	contact := rf.lastContact
+	return contact
+}
+
+func (rf *Raft) getID() (ID int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	ID = rf.me
+	return
+}
+
+func (rf *Raft) getTerm() (term int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	term = rf.currentTerm
+	return
+}
+
```

Comment:
All of the above methods all for safe concurrent access to the Raft's data. Generally getters and setters aren't considered ideomatic `Go` (as far as I've read); however in this case, due to the concurrent access, it was convenient to wrap the mutex locking/unlocking in methods.